src/zenml/integrations/hyperai/orchestrators/hyperai_orchestrator.py

```diff
@@ -0,0 +1,324 @@
+                f"{ENV_ZENML_HYPERAI_RUN_ID}."
+            )
+
+    def _validate_mount_path(self, path) -> str:
+        """Validates if a given string is in a valid path format.
+
+        Args:
+            path: The path to be validated.
+
+        Returns:
+            The path in a valid format.
+
+        Raises:
+            RuntimeError: If the path is not in a valid format.
+        """
+        # Define a regular expression pattern to match a valid path format
+        pattern = r'^(?:[a-zA-Z]:\\(\\[^\\/:*?"<>|]*)*$|^/([^\0]*)*$)'
+
+        if bool(re.match(pattern, path)):
+            return path
+        else:
+            raise RuntimeError(
+                f"Path '{path}' is not in a valid format, so a mount cannot be established."
+            )
+
+    def prepare_or_run_pipeline(
+        self,
+        deployment: "PipelineDeploymentResponse",
+        stack: "Stack",
+        environment: Dict[str, str],
+    ) -> Any:
+        """Sequentially runs all pipeline steps in Docker containers.
+
+        Assumes that:
+        - A HyperAI (hyperai.ai) instance is running on the configured IP address.
+        - The HyperAI instance has been configured to allow SSH connections from the
+            machine running the pipeline.
+        - Docker and Docker Compose are installed on the HyperAI instance.
+        - A key pair has been generated and the public key has been added to the
+            HyperAI instance's `authorized_keys` file.
+        - The private key is available in a HyperAI service connector linked to this
+            orchestrator.
+
+        Args:
+            deployment: The pipeline deployment to prepare or run.
+            stack: The stack the pipeline will run on.
+            environment: Environment variables to set in the orchestration
+                environment.
+
+        Raises:
+            RuntimeError: If a step fails.
+        """
+        # Basic Docker Compose definition
+        compose_definition = {"version": "3", "services": {}}
+
+        # Get deployment id
+        deployment_id = deployment.id
+
+        # Set environment
+        os.environ[ENV_ZENML_HYPERAI_RUN_ID] = str(deployment_id)
+        environment[ENV_ZENML_HYPERAI_RUN_ID] = str(deployment_id)
+
+        # Add each step as a service to the Docker Compose definition
+        logger.info("Preparing pipeline steps for deployment.")
+        for step_name, step in deployment.step_configurations.items():
+            # Get image
+            image = self.get_image(deployment=deployment, step_name=step_name)
+
+            # Get settings
+            step_settings = cast(
+                HyperAIOrchestratorSettings, self.get_settings(step)
+            )
+
+            # Define container name as combination between deployment id and step name
+            container_name = f"{deployment_id}-{step_name}"
+
+            # Make Compose service definition for step
+            compose_definition["services"][container_name] = {
```

Comment:
I personally haven't ran Docker containers with GPU support, but don't we need to set some special options to allow the containers to access the GPUs ? Something like this maybe: https://stackoverflow.com/questions/25185405/using-gpu-from-a-docker-container ?

Reply:
@michael-zenml or is this something that ZenML's docker settings can be used to configure on a per-container level ?

Reply:
@schustmi and me discussed this and we definitely need cuda installed but not sure we need anything else

Reply:
As far as I know:

* The NVIDIA driver for Linux: https://www.nvidia.com/en-us/drivers/unix/
* The NVIDIA Container Toolkit (https://github.com/NVIDIA/nvidia-container-toolkit) must be installed and configured (https://docs.nvidia.com/datacenter/cloud-native/container-toolkit/latest/install-guide.html)

That's it! No CUDA toolkit needs to be installed on the host, it only needs to be available within the container for which many packages are available:

![image](https://github.com/zenml-io/zenml/assets/6450612/9d3c7a9b-e9be-47a5-a293-9f807d161843)

Perhaps we can add a section about running GPU containers and these necessities into the orchestrator docs? I think it's difficult to install these from within the orchestrator or while installing the integration.

Reply:
@schustmi @htahir1 WDYT?

Reply:
I think our `DockerSettings` can be used to configure an image that is GPU enabled, I'm not 100% sure if there's anything else needed in the docker-compose file to make it work though (something like the `--gpus=all` flag in @stefannica link above). If not, I don't think there is any need to mention this in the orchestrator docs as it's a general solution that applies to all orchestrators.

Reply:
Ah, yes. We will probably need to add this to the Docker Compose file generation:

```
    resources:
        reservations:
          devices:
            - driver: nvidia
              capabilities: [gpu]
```

That way, all GPUs should become accessible to the container (because we omit setting `count` and `device_ids`, see link below).

Perhaps, we can make this configurable (i.e. allow orchestrators to use parts of the GPU) later.

https://docs.docker.com/compose/gpu-support/

Reply:
I've just tested on a GPU equipped instance with a local Docker image (`tensorflow:latest-gpu` with ZenML installed locally). It then does run (and not only create) the machine, compared to a CPU-only instance (so the configurable disabling of GPU reservation would be a solution). Within the pipeline, I ran

```
    import tensorflow as tf

    # check physical devices
    logger.info('Physical devices:')
    logger.info(tf.config.list_physical_devices())
```

and got

```
Physical devices:
[PhysicalDevice(name='/physical_device:CPU:0', device_type='CPU'), PhysicalDevice(name='/physical_device:GPU:0', device_type='GPU')]
```

so the way GPU configuration is built into the orchestrator is working, under the condition that the NVIDIA Driver and NVIDIA Container Toolkit are installed on the instance!